File: pocs/code-review-demo/core/event_bridge.py
from typing import Callable, Awaitable
from core.models import ReviewEvent


async def attach_bridge(agent, ws_send: Callable[[str], Awaitable[None]], agent_name: str) -> None:
    @agent.emitter.on("*.*")
    async def handle(data, event):
        review_event = _map_event(event.name, data, agent_name)
        if review_event:
            await ws_send(review_event.to_json())


def _map_event(event_name: str, data, agent_name: str) -> ReviewEvent | None:
    if "think" in event_name:
        return ReviewEvent(type="agent.thinking", agent=agent_name, message=str(data), severity=None)
    if "tool" in event_name and "start" in event_name:
        return ReviewEvent(type="tool.start", agent=agent_name, message=f"Using tool: {event_name}", severity=None)
    if "tool" in event_name and "end" in event_name:
        return ReviewEvent(type="tool.end", agent=agent_name, message=str(data), severity=None)
    return None

File: pocs/code-review-demo-ts/src/core/eventBridge.ts
import { ReviewEvent, makeEvent } from "./models.js";

type WsSend = (event: ReviewEvent) => void;

export function attachBridge(agent: any, send: WsSend, agentName: ReviewEvent["agent"]) {
  agent.emitter.match("*.*", (data: unknown, event: { path: string }) => {
    const mapped = mapEvent(event.path, data, agentName);
    if (mapped) send(mapped);
  });
}

function mapEvent(path: string, data: unknown, agentName: ReviewEvent["agent"]): ReviewEvent | null {
  if (path.includes("think")) {
    return makeEvent("agent.thinking", agentName, extractText(data));
  }
  if (path.includes("tool") && path.includes("start")) {
    return makeEvent("tool.start", agentName, `Using tool: ${path}`);
  }
  if (path.includes("tool") && path.includes("end")) {
    return makeEvent("tool.end", agentName, extractText(data));
  }
  if (path.includes("handoff")) {
    return makeEvent("agent.handoff", agentName, extractText(data));
  }
  if (path.includes("start")) {
    return makeEvent("agent.start", agentName, `Agent ${agentName} started`);
  }
  return null;
}

function extractText(data: unknown): string {
  if (typeof data === "string") return data;
  if (data && typeof data === "object") {
    const d = data as any;
    return d.text ?? d.message ?? d.output ?? JSON.stringify(data);
  }
  return String(data);
}
